src/components/Canvas.tsx

```diff
@@ -1,12 +1,70 @@
 import * as React from "react";
 
+import {Circle} from "./Circle.tsx";
+import {Square} from "./Square.tsx";
+import {Triangle} from "./Triangle.tsx";
+
+import * as models from "../models.ts";
+
 export interface CanvasProps {
 }
 
-export class Canvas extends React.Component<CanvasProps, {}> {
+export interface CanvasState {
+  shapes: models.Shape[];
+}
+
+export class Canvas extends React.Component<CanvasProps, CanvasState> {
+  constructor(props: CanvasProps) {
+    super(props);
+    this.state = {shapes: []};
+  }
+
   render() {
-    return <div className="canvas">
-      <svg></svg>
+    return <div className="canvas" onDrop={(e) => {this.drop(e)}} onDragOver={(e) => {this.dragOver(e)}}> 
+      <svg ref="svg">
+        {
+          this.state.shapes.map((shape) => {
+            return this.renderShape(shape);
+          })
+        }
+      </svg>
     </div>;
   }
+
+  private renderShape(shape: models.Shape) {
+    let key = Math.random();
```

Comment:
Here's the docs on keys and performance: https://facebook.github.io/react/docs/reconciliation.html#keys